src/osd/PrimaryLogPG.cc

```diff
@@ -9976,6 +9978,264 @@ struct C_Flush : public Context {
   }
 };
 
+struct C_SetDedupChunks : public Context {
+  PrimaryLogPGRef pg;
+  hobject_t oid;
+  epoch_t last_peering_reset;
+  ceph_tid_t tid;
+  uint64_t offset;
+  
+  C_SetDedupChunks(PrimaryLogPG *p, hobject_t o, epoch_t lpr)
+    : pg(p), oid(o), last_peering_reset(lpr),
+      tid(0)
+  {}
+  void finish(int r) override {
+    if (r == -ECANCELED)
+      return;
+    std::scoped_lock locker{*pg};
+    auto it = pg->manifest_ops.find(oid);
+    if (it == pg->manifest_ops.end()) {
+      // raced with cancel_manifest_ops
+      return;
+    }
+    // check if the previous op returns fail
+    if (it->second->results[0] < 0) {
+      return;
+    }
+
+    it->second->results[offset] = r;
+    if (last_peering_reset == pg->get_last_peering_reset() &&
```

Comment:
If last_peering_reset != pg->get_last_peering_reset(), then pg->manifest_ops must have been already cleared as part of the peering reset.  Do this check first immediately after taking the lock -- you should simply drop the lock and return in that case without checking any other pg state.

Reply:
See cancel_manifest_ops.